src/string/grapheme/grapheme.cr

```diff
@@ -0,0 +1,226 @@
+require "./properties"
+
+module String::Grapheme
+  # Unicode extended grapheme cluster. `Graphemes` Iterator will return an instance of this struct to return information about specific grapheme cluster.
+  struct Cluster
```

Comment:
This is a great start, thank you! 🙏 

I was thinking... most of the time a cluster will be just a single char, right? I was thinking that a `Cluster` could still be a struct, but the underlying representation could be a union:
- if the cluster is simple (just one codepoint) then it's just that codepoint
- if the cluster consists of multiple codepoints, it's an array of codepoints. Or if clusters of two codepoints are common, we could also consider a `Tuple(Int32, Int32)` to be part another representation (same if three codepoints are common)
- we should probably not keep a reference to the original string. Imagine reading a huge string from a file and keeping the first grapheme. The grapheme will point back to the string so it will never be GCed. So in my mind a grapheme must be an independent entity.

Reply:
Thank you Ary 🙏 for review and comments.

>     * if the cluster is simple (just one codepoint) then it's just that codepoint
Yes, there would be cases where cluster will have just one codepoint, but some emojis or some unicode chars can have 4 or more codepoints. for example `🧙‍♂️` will have 4 codepoints of `[129497, 8205, 9794, 65039]`.


>     * if the cluster consists of multiple codepoints, it's an array of codepoints. Or if clusters of two codepoints are common, we could also consider a `Tuple(Int32, Int32)` to be part another representation (same if three codepoints are common)

As for item 1, there could be clusters with multiple codepoints, so don't know what would be the optimal case for Tuple size? should we think of cases with 2,3,4 and so on?


>     * we should probably not keep a reference to the original string. Imagine reading a huge string from a file and keeping the first grapheme. The grapheme will point back to the string so it will never be GCed. So in my mind a grapheme must be an independent entity.

Code is not keeping reference to original string, but its keeping the codepoints and byte location of chars, and `Graphemes` constructor is the only place, where passed string is handled and then forget of. Please correct me if some portion of code is keeping the reference to original string.